File: platform/polyglot/rag_search_algorithms/mjs/src/post-processor.mjs
/**
 * @fileoverview Post-processing pipeline for search results.
 */

import { separateCodeText } from './code-text-separator.mjs';
import { detectComponentsMetadata, detectComponentsParse } from './component-detector.mjs';

/**
 * Post-process hybrid search results: code/text separation + component detection.
 *
 * @param {Array<[{metadata: Record<string, any>, pageContent?: string, page_content?: string}, number]>} results
 * @param {Object} [opts]
 * @param {string} [opts.codeMode='regex']
 * @param {string} [opts.componentMode='metadata']
 * @param {string[]} [opts.importPackages]
 * @returns {import('./types.mjs').PostProcessedResults}
 */
export function postProcessResults(results, opts = {}) {
  const { componentMode = 'metadata', importPackages } = opts;

  const components = componentMode === 'parse'
    ? detectComponentsParse(results, importPackages)
    : detectComponentsMetadata(results);

  const processed = results.map(([doc, score]) => {
    const content = (doc.pageContent || doc.page_content || '').slice(0, 500);
    const fileName = doc.metadata?.file_name || '';
    const sep = separateCodeText(content, fileName);

    return {
      content,
      codeParts: sep.codeParts,
      textParts: sep.textParts,
      metadata: doc.metadata || {},
      score,
    };
  });

  return { components, results: processed };
}

File: platform/polyglot/rag_search_algorithms/py/rag_search_algorithms/post_processor.py
"""Post-processing pipeline for search results."""

from typing import Optional

from .code_text_separator import separate_code_text_regex
from .component_detector import detect_components_metadata, detect_components_parse


def post_process_results(
    results: list,
    code_mode: str = "regex",
    component_mode: str = "metadata",
    import_packages: Optional[list[str]] = None,
) -> dict:
    """Post-process hybrid search results: code/text separation + component detection.

    Args:
        results: List of (document, score) tuples.
        code_mode: Code separation mode ("regex" for heuristic separation).
        component_mode: Component detection mode ("metadata" or "parse").
        import_packages: Package names for import pattern matching.

    Returns:
        Dict with "components" (list[str]) and "results" (list[dict]).
    """
    components = (
        detect_components_parse(results, import_packages=import_packages)
        if component_mode == "parse"
        else detect_components_metadata(results)
    )

    truncated_contents = [doc.page_content[:500] for doc, _score in results]
    file_names = [doc.metadata.get("file_name", "") for doc, _score in results]

    separations = [
        separate_code_text_regex(c, fn)
        for c, fn in zip(truncated_contents, file_names)
    ]

    processed = []
    for i, (doc, score) in enumerate(results):
        sep = separations[i] if i < len(separations) else {"code_parts": [], "text_parts": []}
        processed.append({
            "content": truncated_contents[i],
            "code_parts": sep["code_parts"],
            "text_parts": sep["text_parts"],
            "metadata": doc.metadata,
            "score": score,
        })

    return {"components": components, "results": processed}
